File: kbe/res/sdk_templates/client/ue4/Source/KBEnginePlugins/Engine/EncryptionFilter.cpp
#include "EncryptionFilter.h"
#include "MemoryStream.h"
#include "MessageReader.h"
#include "PacketSenderTCP.h"
#include "PacketSenderKCP.h"
#include "Engine/KBDebug.h"

#include "cryptlib.h"
#include "rdrand.h"
#include "modes.h"
#include "secblock.h"

BlowfishFilter::BlowfishFilter(int keySize):
	isGood_(false),
	pPacket_(new MemoryStream()),
	pEncryptStream_(new MemoryStream()),
	packetLen_(0),
	padSize_(0)
{
	key_.Init(0, keySize);

	CryptoPP::RDRAND rng;
	rng.GenerateBlock(key_.GetData(), key_.Num());
	init();
}

BlowfishFilter::BlowfishFilter(const TArray<uint8>& key):
	isGood_(false),
	key_(key),
	pPacket_(new MemoryStream()),
	pEncryptStream_(new MemoryStream()),
	packetLen_(0),
	padSize_(0)
{
	init();
}

BlowfishFilter::~BlowfishFilter()
{
	KBE_SAFE_RELEASE(pPacket_);
	KBE_SAFE_RELEASE(pEncryptStream_);
}

bool BlowfishFilter::init()
{
	if (key_.Num() >= encripter.MinKeyLength() && key_.Num() <= encripter.MaxKeyLength())
	{
		encripter.SetKey(key_.GetData(), key_.Num());
		decripter.SetKey(key_.GetData(), key_.Num());
		isGood_ = true;
	}
	else
	{
		ERROR_MSG("BlowfishFilter::init: invalid length %d", key_.Num());
		isGood_ = false;
	}

	return isGood_;
}

void BlowfishFilter::encrypt(MemoryStream *pMemoryStream)
{
	// BlowFish 每次只能加密和解密8字节数据
	// 不足8字节则填充0
	uint8 padSize = 0;

	if (pMemoryStream->length() % BLOCK_SIZE != 0)
	{
		padSize = BLOCK_SIZE - pMemoryStream->length() % BLOCK_SIZE;
		pMemoryStream->data_resize(pMemoryStream->size() + padSize);
		memset(pMemoryStream->data() + pMemoryStream->wpos(), 0, padSize);
		pMemoryStream->wpos(pMemoryStream->wpos() + padSize);
	}

	encrypt(pMemoryStream->data() + pMemoryStream->rpos(), pMemoryStream->length());

	uint16 packetLen = pMemoryStream->length() + 1;
	pEncryptStream_->writeUint16(packetLen);
	pEncryptStream_->writeUint8(padSize);
	pEncryptStream_->append(pMemoryStream->data() + pMemoryStream->rpos(), pMemoryStream->length());

	pMemoryStream->swap(*pEncryptStream_);
	pEncryptStream_->clear(false);
}

void BlowfishFilter::encrypt(uint8 *buf, MessageLengthEx len)
{
	if (len % BLOCK_SIZE != 0)
	{
		ERROR_MSG("BlowfishFilter::encrypt: Input length (%d) is not a multiple of block size ", len);
		return;
	}

	uint8* data = buf;
	uint64 prevBlock = 0;
	for (uint32 i = 0; i < len; i += BLOCK_SIZE)
	{
		if (prevBlock != 0)
		{
			uint64 oldValue = *(uint64*)(data + i);
			*(uint64*)(data + i) = *(uint64*)(data + i) ^ (prevBlock);
			prevBlock = oldValue;
		}
		else
		{
			prevBlock = *(uint64*)(data + i);
		}

		encripter.ProcessData(data + i, data + i, BLOCK_SIZE);
	}
}

void BlowfishFilter::encrypt(uint8 *buf, MessageLengthEx offset, MessageLengthEx len)
{
	encrypt(buf + offset, len);
}

void BlowfishFilter::decrypt(MemoryStream *pMemoryStream)
{
	decrypt(pMemoryStream->data() + pMemoryStream->rpos(), pMemoryStream->length());
}

void BlowfishFilter::decrypt(uint8 *buf, MessageLengthEx len)
{
	if (len % BLOCK_SIZE != 0)
	{
		ERROR_MSG("BlowfishFilter::decrypt: Input length (%d) is not a multiple of block size ", len);
		return;
	}

	uint8* data = buf;
	uint64 prevBlock = 0;
	for (uint32 i = 0; i < len; i += BLOCK_SIZE)
	{
		decripter.ProcessData(data + i, data + i, BLOCK_SIZE);

		if (prevBlock != 0)
		{
			*(uint64*)(data + i) = *(uint64*)(data + i) ^ (prevBlock);
		}
		
		prevBlock = *(uint64*)(data + i);
	}
}

void BlowfishFilter::decrypt(uint8 *buf, MessageLengthEx offset, MessageLengthEx len)
{
	decrypt(buf + offset, len);
}

bool BlowfishFilter::send(PacketSenderBase* pPacketSender, MemoryStream *pPacket)
{
	if (!isGood_)
	{
		ERROR_MSG("BlowfishFilter::send: Dropping packet due to invalid filter");
		return false;
	}

	encrypt(pPacket);

	return pPacketSender->send(pPacket);;
}

bool BlowfishFilter::recv(MessageReader* pMessageReader, MemoryStream *pPacket)
{
	if (!isGood_)
	{
		ERROR_MSG("BlowfishFilter::recv: Dropping packet due to invalid filter");
		return false;
	}

	uint32 oldrpos = pPacket->rpos();
	uint32 len = pPacket->length();
	uint16 packeLen = pPacket->readUint16();

	if ( 0 == pPacket_->length() && len > MIN_PACKET_SIZE && packeLen - 1 == len - 3)
	{
		int padSize = pPacket->readUint8();
		decrypt(pPacket);

		if (pMessageReader)
		{
			pMessageReader->process(pPacket->data() + pPacket->rpos(), 0, pPacket->length() - padSize);
		}

		pPacket->clear(false);
		return true;
	}

	pPacket->rpos(oldrpos);
	pPacket_->append(pPacket->data() + pPacket->rpos(), pPacket->length());
	pPacket->clear(false);

	while (pPacket_->length() > 0)
	{
		uint32 currLen = 0;
		int oldwpos = 0;
		if (packetLen_ <= 0)
		{
			if (pPacket_->length() >= MIN_PACKET_SIZE)
			{
				(*pPacket_) >> packetLen_;
				(*pPacket_) >> padSize_;

				packetLen_ -= 1;

				if (pPacket_->length() > packetLen_)
				{
					currLen = (uint32)(pPacket_->rpos() + packetLen_);
					oldwpos = pPacket_->wpos();
					pPacket_->wpos(currLen);
				}
				else if (pPacket_->length() < packetLen_)
				{
					return false;
				}
			}
			else
			{
				return false;
			}
		}
		else
		{
			if (pPacket_->length() > packetLen_)
			{
				currLen = (uint32)(pPacket_->rpos() + packetLen_);
				oldwpos = pPacket_->wpos();
				pPacket_->wpos(currLen);
			}
			else if (pPacket_->length() < packetLen_)
			{
				return false;
			}
		}

		decrypt(pPacket_);
		pPacket_->wpos(pPacket_->wpos() - padSize_);

		if (pMessageReader)
		{
			pMessageReader->process(pPacket_->data() + pPacket_->rpos(), 0, pPacket_->length());
		}

		if (currLen > 0)
		{
			pPacket_->rpos(currLen);
			pPacket_->wpos(oldwpos);
		}
		else
		{
			pPacket_->clear(false);
		}

		packetLen_ = 0;
		padSize_ = 0;
	}
	
	return true;
}

File: kbe/res/sdk_templates/client/unity/EncryptionFilter.cs
namespace KBEngine
{
    using UnityEngine;
    using System;
    using System.Collections;
    using System.Collections.Generic;

    public abstract class EncryptionFilter
    {
        public abstract void encrypt(MemoryStream stream);

        public abstract void decrypt(MemoryStream stream);
        public abstract void decrypt(byte[] buffer, int startIndex, int length);

        public abstract bool send(PacketSenderBase sender, MemoryStream stream);
        public abstract bool recv(MessageReaderBase reader, byte[] buffer, UInt32 rpos, UInt32 len);
    }


    class BlowfishFilter : EncryptionFilter
    {
        private Blowfish _blowfish = new Blowfish();
        
        private MemoryStream _packet = new MemoryStream();
        MemoryStream _enctyptStrem = new MemoryStream();

        private UINT8 _padSize = 0;
        private UInt16 _packLen = 0;
        const UInt32 BLOCK_SIZE = 8;
        const UInt32 MIN_PACKET_SIZE = sizeof(UInt16) + 1 + BLOCK_SIZE;

        public BlowfishFilter()
        {
        }

        ~BlowfishFilter()
        {
        }

        public  byte[] key()
        {
            return _blowfish.key();
        }

        public override void encrypt(MemoryStream stream)
        {
            int padSize = 0;
            if (stream.length() % BLOCK_SIZE != 0)
            {
                padSize = (int)(BLOCK_SIZE - (stream.length() % BLOCK_SIZE));
                stream.wpos += padSize;

				if(stream.wpos > MemoryStream.BUFFER_MAX)
					Debug.LogError("BlowfishFilter::encrypt: stream.wpos(" + stream.wpos + ") > MemoryStream.BUFFER_MAX(" + MemoryStream.BUFFER_MAX + ")!");
            }

            _blowfish.encipher(stream.data(), (int)stream.length());

            UInt16 packLen = (UInt16)(stream.length() + 1);
            _enctyptStrem.writeUint16(packLen);
            _enctyptStrem.writeUint8((UINT8)padSize);
            _enctyptStrem.append(stream.data(), (UInt32)stream.rpos, stream.length());

            stream.swap(_enctyptStrem);
            _enctyptStrem.clear();
        }

        public override void decrypt(MemoryStream stream)
        {
            _blowfish.decipher(stream.data(), stream.rpos, (int)stream.length());
        }

        public override void decrypt(byte[] buffer, int startIndex, int length)
        {
            _blowfish.decipher(buffer, startIndex, length);
        }

        public override bool send(PacketSenderBase sender, MemoryStream stream)
        {
            if(!_blowfish.isGood())
            {
                Dbg.ERROR_MSG("BlowfishFilter::send: Dropping packet, due to invalid filter");
                return false;
            }

            encrypt(stream);
            return sender.send(stream);
        }

        public override bool recv(MessageReaderBase reader, byte[] buffer, UInt32 rpos, UInt32 len)
        {
            if (!_blowfish.isGood())
            {
                Dbg.ERROR_MSG("BlowfishFilter::recv: Dropping packet, due to invalid filter");
                return false;
            }

            if (_packet.length() == 0 && len >= MIN_PACKET_SIZE && BitConverter.ToUInt16(buffer, (int)rpos) - 1 == len - 3)
            {
                int packLen = BitConverter.ToUInt16(buffer, (int)rpos) - 1;
                int padSize = buffer[rpos + 2];

                decrypt(buffer, (int)(rpos + 3), (int)(len - 3));

                int length = packLen - padSize;
                if (reader != null)
                {
                    reader.process(buffer, rpos + 3, (UInt32)length);
                }

                return true;
            }

            _packet.append(buffer, rpos, len);

            while(_packet.length() > 0)
            {
                UInt32 currLen = 0;
                int oldwpos = 0;
                if (_packLen <= 0)
                {
                    // 如果满足一个最小包则尝试解包, 否则缓存这个包待与下一个包合并然后解包
                    if (_packet.length() >= MIN_PACKET_SIZE)
                    {
                        _packLen = _packet.readUint16();
                        _padSize = _packet.readUint8();

                        _packLen -= 1;

                        if (_packet.length() > _packLen)
                        {
                            currLen = (UInt32)(_packet.rpos + _packLen);
                            oldwpos = _packet.wpos;
                            _packet.wpos = (int)currLen;
                        }
                        else if (_packet.length() < _packLen)
                        {
                            return false;
                        }
                    }
                    else
                    {
                        return false;
                    }
                }
                else
                {
                    // 如果上一次有做过解包行为但包还没有完整则继续处理
                    // 如果包是完整的下面流程会解密， 如果有多余的内容需要将其剪裁出来待与下一个包合并
                    if (_packet.length() > _packLen)
                    {
                        currLen = (UInt32)(_packet.rpos + _packLen);
                        oldwpos = _packet.wpos;
                        _packet.wpos = (int)currLen;
                    }
                    else if (_packet.length() < _packLen)
                    {
                        return false;
                    }
                }

                decrypt(_packet);
                _packet.wpos -= _padSize;

                // 上面的流程能保证wpos之后不会有多余的包
                // 如果有多余的包数据会放在_recvStream
                if (reader != null)
                {
                    reader.process(_packet.data(), (UInt32)_packet.rpos, _packet.length());
                }

                if (currLen > 0)
                {
                    _packet.rpos = (int)currLen;
                    _packet.wpos = oldwpos;
                }
                else
                {
                    _packet.clear();
                }

                _packLen = 0;
                _padSize = 0;
            }
            
            return true;
        }
    }
}

